subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -0,0 +1,2516 @@
+			zoap_header_set_code(out.out_zpkt,
+					     ZOAP_RESPONSE_CODE_INTERNAL_ERROR);
+			r = 0;
+		}
+	}
+
+	return r;
+}
+
+void lwm2m_udp_receive(struct net_context *ctx, struct net_pkt *pkt,
+		       struct zoap_pending *zpendings, int num_zpendings,
+		       struct zoap_reply *zreplies, int num_zreplies,
+		       int (*udp_request_handler)(struct zoap_packet *,
+						  struct zoap_packet *,
+						  struct sockaddr *))
+{
+	struct net_udp_hdr hdr, *udp_hdr;
+	struct zoap_pending *pending;
+	struct zoap_reply *reply;
+	struct zoap_packet response;
+	struct sockaddr from_addr;
+	struct zoap_packet response2;
+	struct net_pkt *pkt2;
+	int header_len, r;
+	const u8_t *token;
+	u8_t tkl;
+
+	udp_hdr = net_udp_get_hdr(pkt, &hdr);
+	if (!udp_hdr) {
+		SYS_LOG_ERR("Invalid UDP data");
+		return;
+	}
+
+	/* Save the from address */
+#if defined(CONFIG_NET_IPV6)
+	if (net_pkt_family(pkt) == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&from_addr)->sin6_addr,
+				&NET_IPV6_HDR(pkt)->src);
+		net_sin6(&from_addr)->sin6_port = udp_hdr->src_port;
+		net_sin6(&from_addr)->sin6_family = AF_INET6;
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+	if (net_pkt_family(pkt) == AF_INET) {
+		net_ipaddr_copy(&net_sin(&from_addr)->sin_addr,
+				&NET_IPV4_HDR(pkt)->src);
+		net_sin(&from_addr)->sin_port = udp_hdr->src_port;
+		net_sin(&from_addr)->sin_family = AF_INET;
+	}
+#endif
+
+	/*
+	 * zoap expects that buffer->data starts at the
+	 * beginning of the CoAP header
+	 */
+	header_len = net_pkt_appdata(pkt) - pkt->frags->data;
+	net_buf_pull(pkt->frags, header_len);
+
+	r = zoap_packet_parse(&response, pkt);
+	if (r < 0) {
+		SYS_LOG_ERR("Invalid data received (err:%d)", r);
+		goto cleanup;
+	}
+
+	token = zoap_header_get_token(&response, &tkl);
+	pending = zoap_pending_received(&response, zpendings, num_zpendings);
+	if (pending) {
+		/* If necessary cancel retransmissions */
+	}
+
+	SYS_LOG_DBG("checking for reply from [%s]",
+		    lwm2m_sprint_ip_addr(&from_addr));
+	reply = zoap_response_received(&response, &from_addr,
+				       zreplies, num_zreplies);
+	if (!reply) {
+		/*
+		 * If no normal response handler is found, then this is
+		 * a new request coming from the server.  Let's look
+		 * at registered objects to find a handler.
+		 */
+		if (udp_request_handler &&
+		    zoap_header_get_type(&response) == ZOAP_TYPE_CON) {
+			/* Create a response packet if we reach this point */
+			r = lwm2m_init_message(ctx, &response2, &pkt2,
+					       ZOAP_TYPE_ACK,
+					       zoap_header_get_code(&response),
+					       zoap_header_get_id(&response),
+					       NULL, -1);
+			if (r < 0) {
+				if (pkt2) {
+					net_pkt_unref(pkt2);
+				}
+				goto cleanup;
+			}
+
+			/*
+			 * The "response" here is actually a new request
+			 */
+			r = udp_request_handler(&response, &response2,
+						&from_addr);
+			if (r < 0) {
+				SYS_LOG_ERR("Request handler error: %d", r);
+			} else {
+				r = net_context_sendto(pkt2, &from_addr,
```

Comment:
BTW, if we want to have transparent DTLS support, then the net-app API net_app_send_pkt() should be used here, instead of direct call to net_context_sendto(). This can be changed in later PR of course.